tests/phpunit/CRM/Core/BAO/SettingTest.php

```diff
@@ -180,4 +180,30 @@ public static function _testOnChange_onChangeExample($oldValue, $newValue, $meta
     $_testOnChange_hookCalls['metadata'] = $metadata;
   }
 
+  /**
+   * Test to set civicrmEnvironment
+   *
+   */
+  public function testSetCivicrmEnvironment() {
+    CRM_Core_BAO_Setting::setItem(TRUE, CRM_Core_BAO_Setting::DEVELOPER_PREFERENCES_NAME, 'civicrmEnvironment');
+    $values = CRM_Core_BAO_Setting::getItem(CRM_Core_BAO_Setting::DEVELOPER_PREFERENCES_NAME, 'civicrmEnvironment');
+    $this->assertEquals(TRUE, $values);
+
+    global $civicrm_setting;
+    $civicrm_setting[CRM_Core_BAO_Setting::DEVELOPER_PREFERENCES_NAME]['civicrmEnvironment'] = FALSE;
+    Civi::service('settings_manager')->useMandatory();
+    $values = CRM_Core_BAO_Setting::getItem(CRM_Core_BAO_Setting::DEVELOPER_PREFERENCES_NAME, 'civicrmEnvironment');
+    $this->assertEquals(FALSE, $values);
+
+    // check for fatal error
+    try {
+      CRM_Core_BAO_Setting::setItem(TRUE, CRM_Core_BAO_Setting::DEVELOPER_PREFERENCES_NAME, 'civicrmEnvironment');
```

Comment:
So tarball should include a commented out line in civicrm.settings.php that would set the value to production=FALSE at end of installation procedure. Let's not try to change the file on upgrade unless there is already a pattern for inserting new code into it on upgrade (I'm pretty sure there is not).

As a result, I'm not sure how to run a test for this error being raised correctly. Can you check if it is possible to open civicrm.settings.php for write/append in the automated test environment, or if it correctly makes that file write-protected? If it is possible, append a line to file that sets the value one way and run test trying to change it and another trying to set it to same value. Make sure that once test is completed that civicrm.settings.php is in original state (this may require copying it temporarily elsewhere, then deleting the changed version and moving copy of original back).